### PR TITLE
Handle nil http request body

### PR DESCRIPTION
**Pull request by jadeallenx**
I'm implementing a Glacier client in go and a lot of the requests have nil bodies. 

This patch fixes a panic in the current aws4 library when the HTTP request has a nil body.  I also added an example test using a nil body to make sure it works correctly.

Thanks.

### CODE REVIEW

**Comment by mikedewar** [comment]
++

**Comment by mikedewar** [comment]
ARGH PLEASE MERGE THIS

**Comment by danp** [comment]
@mrallen1 this could use a rebase and gofmt.

@bmizerany any chance this could get merged?

File: example_test.go [hunk 1]
```diff
@@ -12,7 +12,7 @@ import (
 func ExampleSign() {
 	data := strings.NewReader("{}")
 	r, _ := http.NewRequest("POST", "https://dynamodb.us-east-1.amazonaws.com/", data)
-	r.Header.Set("Host", "dynamodb.us-east-1.awsamazon.com")
```
Comment (bmizerany) [inline example_test.go]: :( Sorry.

File: sign.go [hunk 2]
```diff
@@ -137,11 +137,19 @@ func (s *Service) writeHeaderList(w io.Writer, r *http.Request) {
...
-	}
-	r.Body = ioutil.NopCloser(bytes.NewBuffer(b))
+    var b []byte
+    // If the payload is empty, use the empty string as the input to the SHA256 function
+    // http://docs.amazonwebservices.com/general/latest/gr/sigv4-create-canonical-request.html
+    if r.Body == nil {
+        b = []byte("")
+    } else {
+        var err error
+	    b, err = ioutil.ReadAll(r.Body)
```
Comment (bmizerany) [inline sign.go]: Please run this through gofmt.

**Comment by bmizerany** [comment]
Thank you! I can merge this once you rebase and gofmt.

**Comment by bmizerany** [comment]
I know there are few tests in this project, but do you mind taking a minute to whip one up for this particular fix? I'd like to get some momentum going there. :)

**Comment by jadeallenx** [comment]
I'll try to get a rebase done today (or tonight.)

**Comment by jadeallenx** [comment]
@bmizerany ping - what do you think?

**Comment by jadeallenx** [comment]
@bmizerany OK, think this is pretty legit now. Thanks!

**Comment by jadeallenx** [comment]
It's awesome to finally get this merged :+1: Thanks.
